### ISSUE TITLE
Regression: Attribute self-reference false positive

### ISSUE THREAD

**Issue by InSyncWithFoo** [microsoft/pyright#8398]
Minimal reproducible example ([playground](https://pyright-play.net/?pyrightVersion=1.1.371&pythonVersion=3.13&strict=true&code=MYGwhgzhAEDCBcBYAUNaBLAdgF3tbAngA4CmA2ltgLrQC8GOQA)):

```python
class C:
  int: type[int] = int  # error: Type of "int" could not be determined because it refers to itself
```

This regression was introduced in [v1.1.363](https://pyright-play.net/?pyrightVersion=1.1.363&pythonVersion=3.13&strict=true&code=MYGwhgzhAEDCBcBYAUNaBLAdgF3tbAngA4CmA2ltgLrQC8GOQA) ([v1.1.362](https://pyright-play.net/?pyrightVersion=1.1.362&pythonVersion=3.13&strict=true&code=MYGwhgzhAEDCBcBYAUNaBLAdgF3tbAngA4CmA2ltgLrQC8GOQA) gives the code a pass).

(from [this Stack Overflow question](https://stackoverflow.com/q/78740781))

**Comment by erictraut** [comment]
I'll look at this and see if it's easy to solve, but this behavior isn't something you should rely on. The resolution order for type annotations at runtime is not well documented and can change based on the Python version and on various modes like `from future import __annotations__`. It's best to avoid shadowing the names of builtins in your code. If that's not possible, you should use unambiguous qualified names or type aliases to refer to the type you intend.

**Comment by erictraut** [comment]
Actually, pyright's current behavior is correct in that it mirrors the runtime behavior, so this isn't a bug.

You can see this by running the following code:

```python
class C:
    int: int = 1

print(C.__annotations__) # Prints: {'int': 1}
```

So, you can see that when the runtime evaluates the annotation expression, it's actually using the class-scoped `int` symbol rather than `builtins.int` symbol. That's because the runtime defers evaluation of annotations at runtime for statements of the form `x: y = z` within a class body. Note that if you separate the assignment onto a different line, the behavior changes in both the runtime and in pyright. In both cases, pyright is correctly matching the runtime.

```python
class C:
    int: int
    int = 1

print(C.__annotations__) # Prints {'int': <class 'int'>}
```
